File: backend/drafts.py
import os
import json
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session, select
from groq import Groq, APIError, RateLimitError
import logging

from database import get_session
from auth import get_current_user
from models import User, Draft, LegalNoticeRequest, DraftRead

router = APIRouter(prefix="/drafts", tags=["drafts"])
logger = logging.getLogger(__name__)

# Initialize Groq client
try:
    groq_client = Groq(api_key=os.getenv("GROQ_API_KEY"))
except Exception as e:
    groq_client = None
    logger.error(f"Failed to initialize Groq client: {e}")

GROQ_MODEL = os.getenv("GROQ_MODEL", "llama3-70b-8192")

@router.post("/generate", response_model=DraftRead)
def generate_draft(
    request: LegalNoticeRequest,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    if not groq_client:
        raise HTTPException(status_code=500, detail="Groq API client is not configured.")

    prompt = f"""
You are an expert Indian lawyer. Draft a formal Legal Notice based on the following details.
The legal notice must be drafted strictly in accordance with Indian legal formatting and standards.
Do not include any placeholders that cannot be deduced from the provided information.

Sender Details: {request.sender_details}
Recipient Details: {request.recipient_details}
Subject: {request.subject}
Facts of the matter: {request.facts}
Relief Sought: {request.relief_sought}
Deadline to Respond: {request.deadline}

Produce the final legal notice only. Do not add any introductory or concluding remarks of your own.
"""
    try:
        completion = groq_client.chat.completions.create(
            messages=[
                {"role": "system", "content": "You are a professional Indian lawyer drafting a legal notice."},
                {"role": "user", "content": prompt}
            ],
            model=GROQ_MODEL,
            max_tokens=2500,
            temperature=0.2
        )
        
        choice = completion.choices[0]
        generated_text = choice.message.content
        
        # Check if the output was cut off
        if choice.finish_reason == "length":
            generated_text += "\n\n[WARNING: The generation was cut off due to length limits. Please review and complete manually.]"
            
    except RateLimitError:
        raise HTTPException(status_code=429, detail="Groq API rate limit exceeded. Please try again later.")
    except APIError as e:
        logger.error(f"Groq API Error: {e}")
        raise HTTPException(status_code=502, detail="Error communicating with the Groq API.")
    except Exception as e:
        logger.error(f"Unexpected error: {e}")
        raise HTTPException(status_code=500, detail="An unexpected error occurred during generation.")

    # Save to DB
    db_draft = Draft(
        user_id=current_user.id,
        document_type="Legal Notice",
        input_fields_json=json.dumps(request.model_dump()),
        generated_text=generated_text
    )
    session.add(db_draft)
    session.commit()
    session.refresh(db_draft)
    
    return db_draft

@router.get("", response_model=list[DraftRead])
def get_drafts(
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    drafts = session.exec(select(Draft).where(Draft.user_id == current_user.id).order_by(Draft.created_at.desc())).all()
    return drafts

File: frontend/src/app/actions/drafts.ts
'use server';

import { cookies } from 'next/headers';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export async function generateDraft(formData: FormData) {
  const token = cookies().get('token')?.value;
  if (!token) return { error: 'Unauthorized' };

  const data = {
    sender_details: formData.get('sender_details') as string,
    recipient_details: formData.get('recipient_details') as string,
    subject: formData.get('subject') as string,
    facts: formData.get('facts') as string,
    relief_sought: formData.get('relief_sought') as string,
    deadline: formData.get('deadline') as string,
  };

  try {
    const res = await fetch(`${API_URL}/drafts/generate`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });

    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      return { error: errorData.detail || `Generation failed with status ${res.status}` };
    }

    return await res.json();
  } catch (e) {
    return { error: 'Failed to communicate with the backend API.' };
  }
}

export async function getDrafts() {
  const token = cookies().get('token')?.value;
  if (!token) return { error: 'Unauthorized', drafts: [] };

  try {
    const res = await fetch(`${API_URL}/drafts`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
      cache: 'no-store',
    });

    if (!res.ok) {
      return { error: 'Failed to fetch past drafts', drafts: [] };
    }

    const drafts = await res.json();
    return { drafts };
  } catch (e) {
    return { error: 'Failed to communicate with the backend API.', drafts: [] };
  }
}
